wd_helper.au3

```diff
@@ -2122,3 +2108,20 @@ EndFunc   ;==>__WD_ElementBuildActionString
 Func __WD_ErrHnd()
 
 EndFunc   ;==>__WD_ErrHnd
+; #INTERNAL_USE_ONLY# ===========================================================================================================
+; Name ..........: __WD_JsonButtonAction
+; Description ...: Create JSON string for button actions
+; Syntax ........: __WD_JsonButtonAction($iButton, $sAction)
+; Parameters ....: $iButton             - an integer value.
+;                  $sAction             - a string value.
+; Return values .: Formatted JSON string
+; Author ........: Danp2
+; Modified ......:
+; Remarks .......:
+; Related .......: _WD_ElementActionEx
+; Link ..........:
+; Example .......: No
+; ===============================================================================================================================
+Func __WD_JsonButtonAction($iButton, $sAction)
+	Return ',{"button": ' & $iButton & ', "type": "' & $sAction & '"}'
```

Comment:
@mlipok Do you think there is any value in making the comma at the beginning optional?

Reply:
There was inconsistency in JSON formating before.
for example there was:
```AutoIt
$sPostHoverAction = ',{"button":' & $iButton & ',"type":"pointerDown"},{"type": "pause", "duration": ' & $iHoldDelay & '},{"button":' & $iButton & ',"type":"pointerUp"}'
.....
$sPostAction = ',{"type": "key", "id": "keyboard_2", "actions": [{"type": "keyUp", "value": "' & $sModifier & '"}]}'

```
I just unificate them, but we can stick with a version which will not contain `space` after `:`

Reply:
btw. I just tested example JSON i posted above, by putting them to Notepad++ and Tidy them...
All instances of `:` was alwas suplemented with `space`.

Reply:
But this is trully not necesary and we can unificate them by removing them.
It will have positive impact, less data sent to WebDriver.

Reply:
1) Spaces outside of quotes are optional in JSON, so it's purely an issue of aesthetics
2) My question was regarding the comma at the beginning of the string, not spaces.

Reply:
> 2\. My question was regarding the comma at the beginning of the string, not spaces.

It was desinge this way before:
```AutoIt
		Case 'doubleclick'
			$sPostHoverAction = ',{"button":' & $iButton & ',"type":"pointerDown"},{"button":' & $iButton & ',"type":"pointerUp"},{"button":' & $iButton & ',"type":"pointerDown"},{"button":' & $iButton & ',"type":"pointerUp"}'

		Case 'rightclick'
			$sPostHoverAction = ',{"button":2,"type":"pointerDown"},{"button":2,"type":"pointerUp"}'

		Case 'clickandhold'
			$sPostHoverAction = ',{"button":' & $iButton & ',"type":"pointerDown"},{"type": "pause", "duration": ' & $iHoldDelay & '},{"button":' & $iButton & ',"type":"pointerUp"}'
```

Reply:
It is because of easier way to ADD it to the "TEMPLATE JSON" 

```AutoIt
			; Build dynamic action string
			$sAction = '{"actions":['

			If $sPreAction Then
				$sAction &= $sPreAction
			EndIf

```

Reply:
> 1. Spaces outside of quotes are optional in JSON, so it's purely an issue of aesthetics
Already removed

Reply:
> It was designed this way before:

Agreed. I was simply asking if it would make sense to make the function more flexible by having the comma be optional. For example --

```autoit
Func __WD_JsonButtonAction($iButton, $sAction, $sPreText = Default, $sPostText = Default)

	If $sPreText = Default Then $sPreText = ','
	If $sPostText = Default Then $sPostText = ''

	Return _
			$sPreText & _
			'{' & _
			'	"button": ' & $iButton & _
			'	,"type": "' & $sAction & '"' & _
			'}' & _
			$sPostText
EndFunc   ;==>__WD_JsonButtonAction
 ```

Ultimately, this would only be useful if __WD_JsonButtonAction is called from additional functions that may not need or want the leading comma.

Reply:
I will answer this weekend

Reply:
as so far `__WD_JsonButtonAction` is used here:

```AutoIt
		Case 'doubleclick'
			$sPostHoverAction = _
					__WD_JsonButtonAction($iButton, "pointerDown") & _
					__WD_JsonButtonAction($iButton, "pointerUp") & _
					__WD_JsonButtonAction($iButton, "pointerDown") & _
					__WD_JsonButtonAction($iButton, "pointerUp")

		Case 'rightclick'
			$sPostHoverAction = _
					__WD_JsonButtonAction(2, "pointerDown") & _
					__WD_JsonButtonAction(2, "pointerUp")

		Case 'clickandhold'
			$sPostHoverAction = _
					__WD_JsonButtonAction($iButton, "pointerDown") & _
					', {' & _
					'	"type":"pause"' & _
					'	,"duration":' & $iHoldDelay & _
					'}' & _
					__WD_JsonButtonAction($iButton, "pointerUp")
....................
....................
....................

		Case 'modifierclick'
			; Hold modifier key down
			$sPreAction = _
					'{' & _
					'	"type":"key"' & _
					'	,"id":"keyboard_1"' & _
					'	,"actions":[' & _
					'		{' & _
					'			"type":"keyDown"' & _
					'			,"value":"' & $sModifier & _
					'		}' & _
					'	]' & _
					'},'

			; Perform click
			$sPostHoverAction = _
					__WD_JsonButtonAction($iButton, "pointerDown") & _
					__WD_JsonButtonAction($iButton, "pointerUp")

```

this mean that `__WD_JsonButtonAction`  only affects `$sPostHoverAction`

which is used only here:

```AutoIt
	Local Static $sActionTemplate = StringReplace( _
			'{' & _
			'	"actions":[' & _ ; Open main action
			'		%s' & _ ; %s > $sPreAction
			'		{' & _ ; Start of default "hover" action
			'			"id":"hover"' & _
			'			,"type":"pointer"' & _
			'			,"parameters":{"pointerType":"mouse"}' & _
			'			,"actions":[' & _ ; Open mouse actions
			'				{' & _
			'					"type":"pointerMove"' & _
			'					,"duration":100' & _
			'					,"x":%s' & _ ; %s > $iXOffset
			'					,"y":%s' & _ ; %s > $iYOffset
			'					,"origin":{' & _
			'						"ELEMENT":"%s"' & _ ; %s > $sElement
			'						,"' & $_WD_ELEMENT_ID & '":"%s"' & _ ; %s > $sElement
			'					}' & _
			'				}' & _
			'				%s' & _ ; %s > $sPostHoverAction
			'			]' & _ ; Close mouse actions
			'		}' & _ ; End of default 'hover' action
			'		%s' & _ ; %s > $sPostAction
			'	]' & _ ; Close main action
			'}', @TAB, '')

```
specifically here:

```AutoIt
			'				}' & _
			'				%s' & _ ; %s > $sPostHoverAction
			'			]' & _ ; Close mouse actions

```

Why to you want to complicate it ?

Reply:
Is it just because you don't like the comma at the beginning of the line?

Reply:
In my experience, this is impractical, I know it is more standard-compliant, but not practical.
Therefore, also when creating SQL queries, I put a comma at the beginning of the line.

I've had many cases where when modifying a SQL query or JSON string, I forgot to control "commas" at the end of a line, but since I started using "commas" at the beginning of a line, all problems are a thing of the past.

And this very case shows that it's easier to add commas in front.

Reply:
My most recent thinking is to change `__WD_JsonButtonAction` (an internal function) into `_WD_JsonAction`, which could also be called by external scripts. This is what I have thus far --
```autoit
Func _WD_JsonAction($sAction, $iValue, $sType = "", $sKey = "")
Local $sJSON = ''

	Switch $sAction
		Case 'mouse'
			$sJSON = _
					'{' & _
					'	"button":' & $iValue & _
					'	,"type":"' & $sType & '"' & _
					'}'
		Case 'pause'
			$sJSON = _
					'{' & _
					'	"type":"pause"' & _
					'	,"duration":' & $iValue & _
					'}'

		Case 'key'
			$sJSON = _
					'{' & _
					'	"type":"key"' & _
					'	,"id":"keyboard_' & $iValue & '"' & _
					'	,"actions":[' & _
					'		{' & _
					'			"type":"' & $sType & '"' & _
					'			,"value":"' & $sKey & _
					'		}' & _
					'	]' & _
					'}'

	EndSwitch

	Return $sJSON
EndFunc   ;==>__WD_JsonAction
```

This allows the calling routine to build a proper JSON "action" without concern for the underlying formatting. It doesn't currently insert the leading comma, so that would need to be handled by the calling routine unless I add that as an option.

Your thoughts on this concept?

Reply:
> Why to you want to complicate it ?

That isn't my intent. 

> Is it just because you don't like the comma at the beginning of the line?

It's not that. My goal would be to make it so that this code could be used in more than one routine.

Reply:
> Your thoughts on this concept?

checking

Reply:
Please take a look now